# *pandas* Basics 


The first thing we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [1]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library

In [2]:
df = pd.read_excel('datasets/Canada.xlsx',
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)

Let's view the top 5 rows of the dataset using the `head()` function.


In [3]:
df.head()

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

We can also view the bottom 5 rows of the dataset using the `tail()` function.


In [4]:
df.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

In [8]:
#pd.set_option("display.max_columns", None)
df.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  1981  1982  1983  \
190  South-Eastern Asia  902  Developing regions  1191  1829  2162  3404   
191     Northern Africa  902  Developing regions     0     0     0     0   
192        Western Asia  902  Developing regions     1     2     1     6   
193      Eastern Africa  902  Developing regions    11    17    11     7   
194      Eastern Africa  902  Developing regions    72   114   102    44   

     1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
190  7583  5907  2741  1406  1411  3004  3801  5870  5416  6547  5105  3723   
191     0     0     0     0     0     0     0     1     0     0     0     0   
192     0    18     7    12     7    18     4    18    41    41    39    73   
193    16     9    15    23    44    68    77    69    73    46    51    41   
194    32    29    43    68    99   187   129    94    61    72    78    58   

     1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
190  2462  1752  1631  1419  1803  2117  2291  1713  1816  1852  3153  2574   
191     0     0     0     0     0     0     0     0     0     0     1     0   
192   144   121   141   134   122   181   171   113   124   161   140   122   
193    34    72    34    51    39    78    50    46    56    91    77    71   
194    39    44    43    49    98   110   191   669  1450   615   454   663   

     2008  2009  2010  2011  2012  2013  
190  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0  
192   133   128   211   160   174   217  
193    64    60   102    69    46    59  
194   611   508   494   434   437   407

In [9]:
df

Type    Coverage          OdName  AREA AreaName   REG  \
0    Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1    Immigrants  Foreigners         Albania   908   Europe   925   
2    Immigrants  Foreigners         Algeria   903   Africa   912   
3    Immigrants  Foreigners  American Samoa   909  Oceania   957   
4    Immigrants  Foreigners         Andorra   908   Europe   925   
..          ...         ...             ...   ...      ...   ...   
190  Immigrants  Foreigners        Viet Nam   935     Asia   920   
191  Immigrants  Foreigners  Western Sahara   903   Africa   912   
192  Immigrants  Foreigners           Yemen   935     Asia   922   
193  Immigrants  Foreigners          Zambia   903   Africa   910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa   910   

                RegName  DEV             DevName  1980  1981  1982  1983  \
0         Southern Asia  902  Developing regions    16    39    39    47   
1       Southern Europe  901   Developed regions     1     0     0     0   
2       Northern Africa  902  Developing regions    80    67    71    69   
3             Polynesia  902  Developing regions     0     1     0     0   
4       Southern Europe  901   Developed regions     0     0     0     0   
..                  ...  ...                 ...   ...   ...   ...   ...   
190  South-Eastern Asia  902  Developing regions  1191  1829  2162  3404   
191     Northern Africa  902  Developing regions     0     0     0     0   
192        Western Asia  902  Developing regions     1     2     1     6   
193      Eastern Africa  902  Developing regions    11    17    11     7   
194      Eastern Africa  902  Developing regions    72   114   102    44   

     1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
0      71   340   496   741   828  1076  1028  1378  1170   713   858  1537   
1       0     0     1     2     2     3     3    21    56    96    71    63   
2      63    44    69   132   242   434   491   872   795   717   595  1106   
3       0     0     0     1     0     1     2     0     0     0     0     0   
4       0     0     2     0     0     0     3     0     1     0     0     0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
190  7583  5907  2741  1406  1411  3004  3801  5870  5416  6547  5105  3723   
191     0     0     0     0     0     0     0     1     0     0     0     0   
192     0    18     7    12     7    18     4    18    41    41    39    73   
193    16     9    15    23    44    68    77    69    73    46    51    41   
194    32    29    43    68    99   187   129    94    61    72    78    58   

     1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
0    2212  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652   
1     113   307   574  1264  1816  1602  1021   853  1450  1223   856   702   
2    2054  1842  2292  2389  2867  3418  3406  3072  3616  3626  4807  3623   
3       0     0     0     0     0     0     0     0     0     0     1     0   
4       0     0     2     0     0     1     0     2     0     0     1     1   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
190  2462  1752  1631  1419  1803  2117  2291  1713  1816  1852  3153  2574   
191     0     0     0     0     0     0     0     0     0     0     1     0   
192   144   121   141   134   122   181   171   113   124   161   140   122   
193    34    72    34    51    39    78    50    46    56    91    77    71   
194    39    44    43    49    98   110   191   669  1450   615   454   663   

     2008  2009  2010  2011  2012  2013  
0    2111  1746  1758  2203  2635  2004  
1     560   716   561   539   620   603  
2    4005  5393  4752  4325  3774  4331  
3       0     0     0     0     0     0  
4       0     0     0     0     1     1  
..    ...   ...   ...   ...   ...   ...  
190  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0  
192  

In [10]:
#pd.set_option("display.max_rows", None)
df

Type    Coverage  \
0    Immigrants  Foreigners   
1    Immigrants  Foreigners   
2    Immigrants  Foreigners   
3    Immigrants  Foreigners   
4    Immigrants  Foreigners   
5    Immigrants  Foreigners   
6    Immigrants  Foreigners   
7    Immigrants  Foreigners   
8    Immigrants  Foreigners   
9    Immigrants  Foreigners   
10   Immigrants  Foreigners   
11   Immigrants  Foreigners   
12   Immigrants  Foreigners   
13   Immigrants  Foreigners   
14   Immigrants  Foreigners   
15   Immigrants  Foreigners   
16   Immigrants  Foreigners   
17   Immigrants  Foreigners   
18   Immigrants  Foreigners   
19   Immigrants  Foreigners   
20   Immigrants  Foreigners   
21   Immigrants  Foreigners   
22   Immigrants  Foreigners   
23   Immigrants  Foreigners   
24   Immigrants  Foreigners   
25   Immigrants  Foreigners   
26   Immigrants  Foreigners   
27   Immigrants  Foreigners   
28   Immigrants  Foreigners   
29   Immigrants  Foreigners   
30   Immigrants  Foreigners   
31   Immigrants  Foreigners   
32   Immigrants    Citizens   
33   Immigrants  Foreigners   
34   Immigrants  Foreigners   
35   Immigrants  Foreigners   
36   Immigrants  Foreigners   
37   Immigrants  Foreigners   
38   Immigrants  Foreigners   
39   Immigrants  Foreigners   
40   Immigrants  Foreigners   
41   Immigrants  Foreigners   
42   Immigrants  Foreigners   
43   Immigrants  Foreigners   
44   Immigrants  Foreigners   
45   Immigrants  Foreigners   
46   Immigrants  Foreigners   
47   Immigrants  Foreigners   
48   Immigrants  Foreigners   
49   Immigrants  Foreigners   
50   Immigrants  Foreigners   
51   Immigrants  Foreigners   
52   Immigrants  Foreigners   
53   Immigrants  Foreigners   
54   Immigrants  Foreigners   
55   Immigrants  Foreigners   
56   Immigrants  Foreigners   
57   Immigrants  Foreigners   
58   Immigrants  Foreigners   
59   Immigrants  Foreigners   
60   Immigrants  Foreigners   
61   Immigrants  Foreigners   
62   Immigrants  Foreigners   
63   Immigrants  Foreigners   
64   Immigrants  Foreigners   
65   Immigrants  Foreigners   
66   Immigrants  Foreigners   
67   Immigrants  Foreigners   
68   Immigrants  Foreigners   
69   Immigrants  Foreigners   
70   Immigrants  Foreigners   
71   Immigrants  Foreigners   
72   Immigrants  Foreigners   
73   Immigrants  Foreigners   
74   Immigrants  Foreigners   
75   Immigrants  Foreigners   
76   Immigrants  Foreigners   
77   Immigrants  Foreigners   
78   Immigrants  Foreigners   
79   Immigrants  Foreigners   
80   Immigrants  Foreigners   
81   Immigrants  Foreigners   
82   Immigrants  Foreigners   
83   Immigrants  Foreigners   
84   Immigrants  Foreigners   
85   Immigrants  Foreigners   
86   Immigrants  Foreigners   
87   Immigrants  Foreigners   
88   Immigrants  Foreigners   
89   Immigrants  Foreigners   
90   Immigrants  Foreigners   
91   Immigrants  Foreigners   
92   Immigrants  Foreigners   
93   Immigrants  Foreigners   
94   Immigrants  Foreigners   
95   Immigrants  Foreigners   
96   Immigrants  Foreigners   
97   Immigrants  Foreigners   
98   Immigrants  Foreigners   
99   Immigrants  Foreigners   
100  Immigrants  Foreigners   
101  Immigrants  Foreigners   
102  Immigrants  Foreigners   
103  Immigrants  Foreigners   
104  Immigrants  Foreigners   
105  Immigrants  Foreigners   
106  Immigrants  Foreigners   
107  Immigrants  Foreigners   
108  Immigrants  Foreigners   
109  Immigrants  Foreigners   
110  Immigrants  Foreigners   
111  Immigrants  Foreigners   
112  Immigrants  Foreigners   
113  Immigrants  Foreigners   
114  Immigrants  Foreigners   
115  Immigrants  Foreigners   
116  Immigrants  Foreigners   
117  Immigrants  Foreigners   
118  Immigrants  Foreigners   
119  Immigrants  Foreigners   
120  Immigrants  Foreigners   
121  Immigrants  Foreigners   
122  Immigrants  Foreigners   
123  Immigrants  Foreigners   
124  Immigrants  Foreigners   
125  Immigrants  Foreigners   
126  Immigrants  Foreigners   
127  Immigrants  Foreigners   
128  Immigra

To start by getting basic information about your `dataframe`: 

In [14]:
df.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


In [15]:
df.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

In [16]:
df.index

RangeIndex(start=0, stop=195, step=1)

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [17]:
print(type(df.columns))
print(type(df.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [18]:
df.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [19]:
df.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [20]:
print(type(df.columns.tolist()))
print(type(df.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


In [21]:
# size of dataframe (rows, columns)
df.shape    

(195, 43)

**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [22]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.

print(df.columns)

df.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
print(df.columns)
df.head(2)

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')
Index([  'OdName', 'AreaName',  'RegName',  'DevName',       1980,       1981,
             1982,       1983,       1984,       1985,       1986,       1987,
             1988,       1989,       1990,       1991,       1992,       1993,
             1994,       1995,       1996,       1997,       1998,       1999,
             2000,       2001,       2002,       2003,       2004,       2005,
          

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  \
0    39    47    71   340   496   741   828  1076  1028  1378  1170   713   
1     0     0     0     0     1     2     2     3     3    21    56    96   

   1994  1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  \
0   858  1537  2212  2555  1999  2395  3326  4067  3697  3479  2978  3436   
1    71    63   113   307   574  1264  1816  1602  1021   853  1450  1223   

   2006  2007  2008  2009  2010  2011  2012  2013  
0  3009  2652  2111  1746  1758  2203  2635  2004  
1   856   702   560   716   561   539   620   603

Let's rename the columns so that they make sense. We can use `rename()` method by passing in a dictionary of old and new names as follows:


In [23]:
df.rename(columns={'OdName':'Country', 
                   'AreaName':'Continent', 
                   'RegName':'Region'}, 
          inplace=True #replace the orginal df
         )
df.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [27]:
print(df.iloc[:1,4:5]) # row starting 0 : 1 ,  column starting 4:5 (not inclueded)

   1980
0    16


In [26]:
df['Total'] = df.iloc[:, 4:].sum(axis=1) #axis=0 is column; axis=1 is row. Here sum of 'row'
df

Country  \
0                                          Afghanistan   
1                                              Albania   
2                                              Algeria   
3                                       American Samoa   
4                                              Andorra   
5                                               Angola   
6                                  Antigua and Barbuda   
7                                            Argentina   
8                                              Armenia   
9                                            Australia   
10                                             Austria   
11                                          Azerbaijan   
12                                             Bahamas   
13                                             Bahrain   
14                                          Bangladesh   
15                                            Barbados   
16                                             Belarus   
17                                             Belgium   
18                                              Belize   
19                                               Benin   
20                                              Bhutan   
21                    Bolivia (Plurinational State of)   
22                              Bosnia and Herzegovina   
23                                            Botswana   
24                                              Brazil   
25                                   Brunei Darussalam   
26                                            Bulgaria   
27                                        Burkina Faso   
28                                             Burundi   
29                                          Cabo Verde   
30                                            Cambodia   
31                                            Cameroon   
32                                              Canada   
33                            Central African Republic   
34                                                Chad   
35                                               Chile   
36                                               China   
37      China, Hong Kong Special Administrative Region   
38          China, Macao Special Administrative Region   
39                                            Colombia   
40                                             Comoros   
41                                               Congo   
42                                          Costa Rica   
43                                       Côte d'Ivoire   
44                                             Croatia   
45                                                Cuba   
46                                              Cyprus   
47                                      Czech Republic   
48               Democratic People's Republic of Korea   
49                    Democratic Republic of the Congo   
50                                             Denmark   
51                                            Djibouti   
52                                            Dominica   
53                                  Dominican Republic   
54                                             Ecuador   
55                                               Egypt   
56                                         El Salvador   
57                                   Equatorial Guinea   
58                                             Eritrea   
59                                             Estonia   
60                                            Ethiopia   
61                                                Fiji   
62                                             Finland   
63                                              France   
64                                               Gabon   
65                                              Gambia   
66                                             Georgia   
67                                             Germany   
68                                           

We can check to see how many null objects we have in the dataset as follows:


In [15]:
df.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


In [28]:
df.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

               1990          1991          1992          1993          1994  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     964.379487   1064.148718   1136.856410   1138.712821    993.153846   
std     3158.730195   2952.093731   3330.083742   3495.220063   3613.336444   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000      3.000000      6.500000     11.500000      9.500000   
50%       38.000000     51.000000     74.000000     85.000000     76.000000   
75%      612.500000    657.500000    655.000000    722.500000    545.000000   
max    31668.000000  23380.000000  34123.000000  33720.000000  39231.000000   

               1995          1996          1997          1998          1999  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     962.625641   1026.076923    989.153846    824.241026    922.143590   
std     3091.492343   3321.045004   3070.761447   2385.943695   2887.632585   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       10.500000     14.500000     19.500000     15.000000     16.000000   
50%       91.000000    118.000000    114.000000    106.000000    116.000000   
75%      550.500000    603.500000    612.500000    535.500000    548.500000   
max    30145.000000  29322.000000  22965.000000  21049.000000  30069.000000   

               2000          2001          2002          2003          2004  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1111.343590   1244.323077   1144.158974   1114.343590   1190.169231   
std     3664.042361   3961.621410   3660.579836   3623.509519   3710.505369   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       16.000000     22.000000     18.500000     21.500000     19.000000   
50%      138.000000    169.000000    165.000000    161.000000    191.000000   
75%      659.000000    793.500000    686.000000    673.500000    756.500000   
max    35529.000000  36434.000000  31961.000000  36439.000000  36619.000000   

               2005          2006          2007          2008          2009  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1320.292308   1266.958974   1191.820513   1246.394872   1275.733333   
std     4425.957828   3926.717747   3443.542409   3694.573544   3829.630424   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       28.500000     25.000000     31.000000     31.000000     36.000000   

##   Indexing and Selection (slicing) - *pandas*


### Select Column

**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.

```python
    df.columnName               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python
    df[['column 1', 'column 2']]  # returns dataframe; see  '[[' & ']]' 
```

***


Example: Let's try filtering on the list of countries ('Country').


In [33]:
df.Country  # returns a series


0                                            Afghanistan
1                                                Albania
2                                                Algeria
3                                         American Samoa
4                                                Andorra
5                                                 Angola
6                                    Antigua and Barbuda
7                                              Argentina
8                                                Armenia
9                                              Australia
10                                               Austria
11                                            Azerbaijan
12                                               Bahamas
13                                               Bahrain
14                                            Bangladesh
15                                              Barbados
16                                               Belarus
17                             

In [35]:
type(df.Country)

pandas.core.series.Series

In [36]:
df['Country']

0                                            Afghanistan
1                                                Albania
2                                                Algeria
3                                         American Samoa
4                                                Andorra
5                                                 Angola
6                                    Antigua and Barbuda
7                                              Argentina
8                                                Armenia
9                                              Australia
10                                               Austria
11                                            Azerbaijan
12                                               Bahamas
13                                               Bahrain
14                                            Bangladesh
15                                              Barbados
16                                               Belarus
17                             

In [37]:
type(df['Country'])

pandas.core.series.Series

In [38]:
df[['Country']] #dataframe 

Country
0                                          Afghanistan
1                                              Albania
2                                              Algeria
3                                       American Samoa
4                                              Andorra
5                                               Angola
6                                  Antigua and Barbuda
7                                            Argentina
8                                              Armenia
9                                            Australia
10                                             Austria
11                                          Azerbaijan
12                                             Bahamas
13                                             Bahrain
14                                          Bangladesh
15                                            Barbados
16                                             Belarus
17                                             Belgium
18                                              Belize
19                                               Benin
20                                              Bhutan
21                    Bolivia (Plurinational State of)
22                              Bosnia and Herzegovina
23                                            Botswana
24                                              Brazil
25                                   Brunei Darussalam
26                                            Bulgaria
27                                        Burkina Faso
28                                             Burundi
29                                          Cabo Verde
30                                            Cambodia
31                                            Cameroon
32                                              Canada
33                            Central African Republic
34                                                Chad
35                                               Chile
36                                               China
37      China, Hong Kong Special Administrative Region
38          China, Macao Special Administrative Region
39                                            Colombia
40                                             Comoros
41                                               Congo
42                                          Costa Rica
43                                       Côte d'Ivoire
44                                             Croatia
45                                                Cuba
46                                              Cyprus
47                                      Czech Republic
48               Democratic People's Republic of Korea
49                    Democratic Republic of the Congo
50                                             Denmark
51                                            Djibouti
52                                            Dominica
53                                  Dominican Republic
54                                             Ecuador
55                                               Egypt
56                                         El Salvador
57                                   Equatorial Guinea
58                                             Eritrea
59                                             Estonia
60                                            Ethiopia
61                                                Fiji
62                                             Finland
63                                              France
64                                               Gabon
65                                              Gambia
66                                             Georgia
67                                             Germany
68                                               Ghana
69                                              Greece
70                                             Grenada
71                                           Guatemala
72                              

In [39]:
type(df[['Country']])

pandas.core.frame.DataFrame

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [20]:
df[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


In [40]:
df.head()

Country Continent           Region             DevName  1980  1981  \
0     Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1         Albania    Europe  Southern Europe   Developed regions     1     0   
2         Algeria    Africa  Northern Africa  Developing regions    80    67   
3  American Samoa   Oceania        Polynesia  Developing regions     0     1   
4         Andorra    Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  \
0    39    47    71   340   496   741   828  1076  1028  1378  1170   713   
1     0     0     0     0     1     2     2     3     3    21    56    96   
2    71    69    63    44    69   132   242   434   491   872   795   717   
3     0     0     0     0     0     1     0     1     2     0     0     0   
4     0     0     0     0     2     0     0     0     3     0     1     0   

   1994  1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  \
0   858  1537  2212  2555  1999  2395  3326  4067  3697  3479  2978  3436   
1    71    63   113   307   574  1264  1816  1602  1021   853  1450  1223   
2   595  1106  2054  1842  2292  2389  2867  3418  3406  3072  3616  3626   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     2     0     0     1     0     2     0     0   

   2006  2007  2008  2009  2010  2011  2012  2013   Total  
0  3009  2652  2111  1746  1758  2203  2635  2004  117278  
1   856   702   560   716   561   539   620   603   31398  
2  4807  3623  4005  5393  4752  4325  3774  4331  138878  
3     1     0     0     0     0     0     0     0      12  
4     1     1     0     0     0     0     1     1      30

Notice that the default index of the dataset is a numeric range from 0 to 194. 

This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [41]:
df.set_index('Country', inplace=True) #inplace=True - repalcing the orginal dataset


# tip: The opposite of set is reset. So to reset the index, we can use df.reset_index()

In [42]:
df.head()

Continent  \
Country                                                                               
Afghanistan                                                                    Asia   
Albania                                                                      Europe   
Algeria                                                                      Africa   
American Samoa                                                              Oceania   
Andorra                                                                      Europe   
Angola                                                                       Africa   
Antigua and Barbuda                                 Latin America and the Caribbean   
Argentina                                           Latin America and the Caribbean   
Armenia                                                                        Asia   
Australia                                                                   Oceania   
Austria                                                                      Europe   
Azerbaijan                                                                     Asia   
Bahamas                                             Latin America and the Caribbean   
Bahrain                                                                        Asia   
Bangladesh                                                                     Asia   
Barbados                                            Latin America and the Caribbean   
Belarus                                                                      Europe   
Belgium                                                                      Europe   
Belize                                              Latin America and the Caribbean   
Benin                                                                        Africa   
Bhutan                                                                         Asia   
Bolivia (Plurinational State of)                    Latin America and the Caribbean   
Bosnia and Herzegovina                                                       Europe   
Botswana                                                                     Africa   
Brazil                                              Latin America and the Caribbean   
Brunei Darussalam                                                              Asia   
Bulgaria                                                                     Europe   
Burkina Faso                                                                 Africa   
Burundi                                                                      Africa   
Cabo Verde                                                                   Africa   
Cambodia                                                                       Asia   
Cameroon                                                                     Africa   
Canada                                                             Northern America   
Central African Republic                                                     Africa   
Chad                                                                         Africa   
Chile                                               Latin America and the Caribbean   
China                                                                          Asia   
China, Hong Kong Special Administrative Region                                 Asia   
China, Macao Special Administrative Region                                     Asia   
Colombia                                            Latin America and the Caribbean   
Comoros                                                                      Africa   
Congo                                                                        Africa   
Costa Rica                                          Latin America and the Caribbean   
Côte d'Ivoire                                                                Africa   
Croatia                                                                 

In [44]:
df=df.reset_index()

In [45]:
df.head()

Country Continent           Region             DevName  1980  1981  \
0     Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1         Albania    Europe  Southern Europe   Developed regions     1     0   
2         Algeria    Africa  Northern Africa  Developing regions    80    67   
3  American Samoa   Oceania        Polynesia  Developing regions     0     1   
4         Andorra    Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  \
0    39    47    71   340   496   741   828  1076  1028  1378  1170   713   
1     0     0     0     0     1     2     2     3     3    21    56    96   
2    71    69    63    44    69   132   242   434   491   872   795   717   
3     0     0     0     0     0     1     0     1     2     0     0     0   
4     0     0     0     0     2     0     0     0     3     0     1     0   

   1994  1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  \
0   858  1537  2212  2555  1999  2395  3326  4067  3697  3479  2978  3436   
1    71    63   113   307   574  1264  1816  1602  1021   853  1450  1223   
2   595  1106  2054  1842  2292  2389  2867  3418  3406  3072  3616  3626   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     2     0     0     1     0     2     0     0   

   2006  2007  2008  2009  2010  2011  2012  2013   Total  
0  3009  2652  2111  1746  1758  2203  2635  2004  117278  
1   856   702   560   716   561   539   620   603   31398  
2  4807  3623  4005  5393  4752  4325  3774  4331  138878  
3     1     0     0     0     0     0     0     0      12  
4     1     1     0     0     0     0     1     1      30

In [49]:
df.head(3)

Country Continent           Region             DevName  1980  1981  \
0  Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1      Albania    Europe  Southern Europe   Developed regions     1     0   
2      Algeria    Africa  Northern Africa  Developing regions    80    67   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   

   2012  2013   Total  
0  2635  2004  410473  
1   620   603  109893  
2  3774  4331  486073  

[3 rows x 39 columns]

In [46]:
df.set_index('Country', inplace=True)

In [47]:
df

Continent  \
Country                                                                               
Afghanistan                                                                    Asia   
Albania                                                                      Europe   
Algeria                                                                      Africa   
American Samoa                                                              Oceania   
Andorra                                                                      Europe   
Angola                                                                       Africa   
Antigua and Barbuda                                 Latin America and the Caribbean   
Argentina                                           Latin America and the Caribbean   
Armenia                                                                        Asia   
Australia                                                                   Oceania   
Austria                                                                      Europe   
Azerbaijan                                                                     Asia   
Bahamas                                             Latin America and the Caribbean   
Bahrain                                                                        Asia   
Bangladesh                                                                     Asia   
Barbados                                            Latin America and the Caribbean   
Belarus                                                                      Europe   
Belgium                                                                      Europe   
Belize                                              Latin America and the Caribbean   
Benin                                                                        Africa   
Bhutan                                                                         Asia   
Bolivia (Plurinational State of)                    Latin America and the Caribbean   
Bosnia and Herzegovina                                                       Europe   
Botswana                                                                     Africa   
Brazil                                              Latin America and the Caribbean   
Brunei Darussalam                                                              Asia   
Bulgaria                                                                     Europe   
Burkina Faso                                                                 Africa   
Burundi                                                                      Africa   
Cabo Verde                                                                   Africa   
Cambodia                                                                       Asia   
Cameroon                                                                     Africa   
Canada                                                             Northern America   
Central African Republic                                                     Africa   
Chad                                                                         Africa   
Chile                                               Latin America and the Caribbean   
China                                                                          Asia   
China, Hong Kong Special Administrative Region                                 Asia   
China, Macao Special Administrative Region                                     Asia   
Colombia                                            Latin America and the Caribbean   
Comoros                                                                      Africa   
Congo                                                                        Africa   
Costa Rica                                          Latin America and the Caribbean   
Côte d'Ivoire                                                                Africa   
Croatia                                                                 

In [48]:
# optional: to remove the name of the index
df.index.name = None

In [49]:
df

Continent  \
Afghanistan                                                                    Asia   
Albania                                                                      Europe   
Algeria                                                                      Africa   
American Samoa                                                              Oceania   
Andorra                                                                      Europe   
Angola                                                                       Africa   
Antigua and Barbuda                                 Latin America and the Caribbean   
Argentina                                           Latin America and the Caribbean   
Armenia                                                                        Asia   
Australia                                                                   Oceania   
Austria                                                                      Europe   
Azerbaijan                                                                     Asia   
Bahamas                                             Latin America and the Caribbean   
Bahrain                                                                        Asia   
Bangladesh                                                                     Asia   
Barbados                                            Latin America and the Caribbean   
Belarus                                                                      Europe   
Belgium                                                                      Europe   
Belize                                              Latin America and the Caribbean   
Benin                                                                        Africa   
Bhutan                                                                         Asia   
Bolivia (Plurinational State of)                    Latin America and the Caribbean   
Bosnia and Herzegovina                                                       Europe   
Botswana                                                                     Africa   
Brazil                                              Latin America and the Caribbean   
Brunei Darussalam                                                              Asia   
Bulgaria                                                                     Europe   
Burkina Faso                                                                 Africa   
Burundi                                                                      Africa   
Cabo Verde                                                                   Africa   
Cambodia                                                                       Asia   
Cameroon                                                                     Africa   
Canada                                                             Northern America   
Central African Republic                                                     Africa   
Chad                                                                         Africa   
Chile                                               Latin America and the Caribbean   
China                                                                          Asia   
China, Hong Kong Special Administrative Region                                 Asia   
China, Macao Special Administrative Region                                     Asia   
Colombia                                            Latin America and the Caribbean   
Comoros                                                                      Africa   
Congo                                                                        Africa   
Costa Rica                                          Latin America and the Caribbean   
Côte d'Ivoire                                                                Africa   
Croatia                                                                      Europe   
Cuba                                                Latin America and th

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:

1. The full row data (all columns)
2. For year 2013
3. For years 1980 to 1985


In [51]:
# 1. the full row data (all columns)

df.loc['Japan'] #returns list

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [52]:
# 1. the full row data (all columns)

df.loc[['Japan']] #returns dataframe

Continent        Region            DevName  1980  1981  1982  1983  \
Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

       1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
Japan   246   198   248   422   324   494   379   506   605   907   956   826   

       1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
Japan   994   924   897  1083  1010  1092   806   817   973  1067  1212  1250   

       2008  2009  2010  2011  2012  2013  Total  
Japan  1284  1194  1168  1265  1214   982  55414

In [53]:
# alternate methods
df.iloc[87]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [55]:
type(df.iloc[87])

pandas.core.series.Series

In [54]:
df.iloc[[87]] #dataframe

Continent        Region            DevName  1980  1981  1982  1983  \
Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

       1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
Japan   246   198   248   422   324   494   379   506   605   907   956   826   

       1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
Japan   994   924   897  1083  1010  1092   806   817   973  1067  1212  1250   

       2008  2009  2010  2011  2012  2013  Total  
Japan  1284  1194  1168  1265  1214   982  55414

In [56]:
type(df.iloc[[87]])

pandas.core.frame.DataFrame

In [58]:
df[df.index == 'Japan']

Continent        Region            DevName  1980  1981  1982  1983  \
Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

       1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  2011  2012  \
Japan   246   198   248  ...  1067  1212  1250  1284  1194  1168  1265  1214   

       2013   Total  
Japan   982  193949  

[1 rows x 38 columns]

In [59]:
# 2. for year 2013

df.loc['Japan', 2013]

982

In [61]:
# alternate 
# year 2013 is the last column, with a positional index of 36

df.iloc[87, 36]

982

In [57]:
# 3. for years 1980 to 1985

df.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]]

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object

In [58]:
fav_years = [1980, 1981, 1982, 1983, 1984, 1984]
df.loc['Japan',fav_years]

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object

In [63]:
# Alternative Method
df.iloc[87, [3, 4, 5, 6, 7, 8]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

Column names that are integers (such as the years) might introduce some confusion. 

For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index.

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [71]:
df.columns = list(map(str, df.columns))

#[print (type(x)) for x in df.columns.values]

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [72]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

## Filtering based on a criteria

To filter the dataframe based on a condition, we simply pass the condition as a boolean vector.

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [59]:
# 1. create the condition boolean series
condition = df['Continent'] == 'Asia'

print(condition)

Afghanistan                                              True
Albania                                                 False
Algeria                                                 False
American Samoa                                          False
Andorra                                                 False
Angola                                                  False
Antigua and Barbuda                                     False
Argentina                                               False
Armenia                                                  True
Australia                                               False
Austria                                                 False
Azerbaijan                                               True
Bahamas                                                 False
Bahrain                                                  True
Bangladesh                                               True
Barbados                                                False
Belarus 

In [60]:
# 2. pass this condition into the dataFrame
df[condition]

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

In [76]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df[(df['Continent']=='Asia') & (df['Region']=='Southern Asia')]


Continent         Region             DevName  1980  \
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013    Total  
Afghanistan                  2004   410473  
Bangladesh                   3789   458976  
Bhutan                        487    41132  
India                       33087  4843328  
Iran (Islamic Republic of)  11291  1231461  
Maldives                        1      210  
Nepal                        1308    71554  
Pakistan                    12603  1691200  
Sri Lanka                    2394  1038506  

[9 rows x 38 columns]

In [78]:
print('data dimensions:', df.shape)
print(df.columns)
df.head(2)

data dimensions: (195, 38)
Index(['Continent', 'Region', 'DevName', '1980', '1981', '1982', '1983',
       '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992',
       '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001',
       '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010',
       '2011', '2012', '2013', 'Total'],
      dtype='object')


Continent           Region             DevName  1980  1981  1982  \
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   

             2011  2012  2013   Total  
Afghanistan  2203  2635  2004  410473  
Albania       539   620   603  109893  

[2 rows x 38 columns]